### PR TITLE
Price change signal add more options

### PR DESCRIPTION
**Pull request by tspenov**
### Example settings structure for `price_absolute_change`

```json
//price >= 0.51
{
  "type": "price_absolute_change",
  "target": {"slug": "santiment"},
  "channel": "telegram",
  "operation": {"above": 0.51}
}
```

```json
// price <= 0.50
{
  "type": "price_absolute_change",
  "target": {"slug": "santiment"},
  "channel": "telegram",
  "operation": {"below": 0.50}
}
```

```json
// price >= 0.49 and price <= 0.51
{
  "type": "price_absolute_change",
  "target": {"slug": "santiment"},
  "channel": "telegram",
  "operation": {"inside_channel": [0.49, 0.51]}
}
```

```json
// price <= 0.49 and price >= 0.51
{
  "type": "price_absolute_change",
  "target": {"slug": "santiment"},
  "channel": "telegram",
  "operation": {"outside_channel": [0.49, 0.51]}
}
```

### Example settings structure for `price_percent_change`

```json
{
  "type": "price_percent_change",
  "target": {"slug": "santiment"},
  "channel": "telegram",
  "time_window": "1d",
  "operation": {"percent_up": 1.0}
}
```

```json
{
  "type": "price_percent_change",
  "target": {"slug": "santiment"},
  "channel": "telegram",
  "time_window": "1d",
  "operation": {"percent_down": 1.0}
}
```

### CODE REVIEW

**Comment by tspenov** [comment]
@Partysun added `move_up`, `move_down` for the price percent - you can see the description.

File: docs/user-triggers-api.md [hunk 1]
```diff
@@ -44,7 +74,17 @@ These are the fields describing a trigger.
   "target": {"slug": "santiment"},
   "channel": "telegram",
   "time_window": "1d",
-  "percent_threshold": 1.0
+  "operation": {"move_up": 1.0}
```
Comment (IvanIvanoff) [inline docs/user-triggers-api.md]: would `percent_up` and `percent_down` be better?

Comment (tspenov) [reply docs/user-triggers-api.md]: Sounds also good. If you have strong preference for these 2 - can change ?

File: lib/sanbase/signals/history/utils.ex [hunk 2]
```diff
@@ -39,4 +40,13 @@ defmodule Sanbase.Signals.History.Utils do
   def average(values) do
     Float.round(Enum.sum(values) / length(values), 2)
   end
+
+  defp evaluate_percent_change(percent_change, percent_threshold)
```
Comment (IvanIvanoff) [inline lib/sanbase/signals/history/utils.ex]: `percent_change_triggered?`. `evaluate` implies that some result is returned while it's only boolean

File: lib/sanbase/signals/trigger/validation.ex [hunk 3]
```diff
@@ -1,21 +1,49 @@
 defmodule Sanbase.Signals.Validation do
   @notification_channels ["telegram", "email"]
 
+  defguard is_valid_price(price) when is_number(price) and price >= 0
+  defguard is_valid_percent(percent) when is_number(percent) and percent >= -100
+  defguard is_valid_percent_change_param(percent) when is_number(percent) and percent > 0
```
Comment (IvanIvanoff) [inline lib/sanbase/signals/trigger/validation.ex]: no `param` at the end

File: lib/sanbase/signals/trigger/validation.ex [hunk 4]
```diff
@@ -1,21 +1,49 @@
...
+      do: :ok
+
+  def valid_percent_operation?(operation),
+    do: {:error, "#{inspect(operation)} is not a valid percent operation"}
+
+  def valid_operation?(%{above: above}) when is_valid_price(above), do: :ok
+  def valid_operation?(%{below: below}) when is_valid_price(below), do: :ok
+
+  def valid_operation?(%{inside_channel: inside_channel} = operation)
+      when is_list_of_two(inside_channel),
```
Comment (IvanIvanoff) [inline lib/sanbase/signals/trigger/validation.ex]: This is an overkill. Also it uses `length` which won't stop counting if you pass a list of 10000 elements. I suggest you just pattern match on the list here like: `valid_operation?(%{inside_channel: [_,_]})`

File: test/sanbase/signals/history/utils_test.exs [hunk 5]
```diff
@@ -12,13 +12,19 @@ defmodule Sanbase.Signals.TriggerHistoryUtilsTest do
       Utils.percent_change_calculations_with_cooldown(percent_changes, percent_threshold, 2)
 
     assert percent_change_calculations |> length() == 8
-    assert filter_with_bigger_threshold(percent_change_calculations) |> length() == 2
+    assert filter_with_bigger_threshold(percent_change_calculations) |> length() == 3
```
Comment (IvanIvanoff) [inline test/sanbase/signals/history/utils_test.exs]: Why without changing the test dat the length changed?

File: test/sanbase/signals/history/utils_test.exs [hunk 6]
```diff
@@ -12,13 +12,19 @@ defmodule Sanbase.Signals.TriggerHistoryUtilsTest do
...
 
     # no cooldown
     percent_change_calculations =
       Utils.percent_change_calculations_with_cooldown(percent_changes, percent_threshold, 0)
 
-    assert filter_with_bigger_threshold(percent_change_calculations) |> length() == 5
+    assert filter_with_bigger_threshold(percent_change_calculations) |> length() == 6
```
Comment (IvanIvanoff) [inline test/sanbase/signals/history/utils_test.exs]: Again why without changing the test dat the length changed?

File: test/sanbase/signals/trigger_evaluator_price_test.exs [hunk 7]
```diff
@@ -36,56 +36,161 @@ defmodule Sanbase.Signals.EvaluatorPriceTest do
       main_contract_address: "0x7c5a0ce9267ed19b22f8cae653f198e3e8daf098"
     })
 
-    # trigger 1: 10% change last 6 hours
-    # trigger 2: above 60 or below 50
-    {trigger1, trigger2, trigger3, trigger4} = setup_triggers(user)
+    # # trigger 1: 10% change last 6 hours
+    # # trigger 2: above 60 or below 50
+    # {trigger1, trigger2, trigger3, trigger4} = setup_triggers(user)
```
Comment (IvanIvanoff) [inline test/sanbase/signals/trigger_evaluator_price_test.exs]: why commented?

File: lib/sanbase/signals/history/utils.ex [hunk 8]
```diff
@@ -6,26 +6,27 @@ defmodule Sanbase.Signals.History.Utils do
...
   * Takes a list of tuples: {grouped_value, current}
   * calculates the percent_change between grouped_value and current
-  * then calculates a list of tuples {percent_change, condition_met} where condition_met is 
+  * then calculates a list of tuples {percent_change, condition_met} where condition_met is
   the percent_change bigger than threshold param and not in cooldown
   """
   @spec percent_change_calculations_with_cooldown(
           list({number(), number()}),
-          float(),
+          float() | map(),
```
Comment (IvanIvanoff) [inline lib/sanbase/signals/history/utils.ex]: is the `float()` Typespec still valid?

File: test/sanbase/signals/history/utils_test.exs (hunk 5 above)
Comment (tspenov) [reply test/sanbase/signals/history/utils_test.exs]: It was using `percent_change > percent_threshold`: https://github.com/santiment/sanbase2/pull/1097/files#diff-27b8ebcbb41dd9e4eda15826cdaf8c5cL24 I think `>=` is better fit (like abov, below, inside, outside - all include `=`). There is test value that is equal, hence the change in test.

File: test/sanbase/signals/history/utils_test.exs (hunk 6 above)
Comment (tspenov) [reply test/sanbase/signals/history/utils_test.exs]: Same as above

File: lib/sanbase/signals/history/utils.ex (hunk 8 above)
Comment (tspenov) [reply lib/sanbase/signals/history/utils.ex]: yes - this function is used from daily_active_addresses historical points also

File: test/sanbase/signals/trigger_evaluator_price_test.exs (hunk 7 above)
Comment (tspenov) [reply test/sanbase/signals/trigger_evaluator_price_test.exs]: removed

**Pull request merged**